### PR TITLE
Add support for "fallback" or "default" encoding for SGFs with no "CA"?

### PR DESCRIPTION
**Pull request by Carton**
Hi,
This is an intial patch for feature request. As mentioned in the commit, some servers generate SGF files with no "CA" property. (In my case, it's a SGF file with player name in GBK encoding, but has no "CA").

I know it's better to be fixed from server side, but I think it would be nice if we can provide some "fallback" encoding for these sgf files rather than using the default "ISO-8859-1" encoding.

3 solutions I can think of:
1. Use "chardet" etc. lib to detect encoding if no "CA" specified.
2. Add extra "fallback_encoding" parameter as in my patch, and config it in config.json (or even better, UI settings).
3. Or if #2 is too confusing, use the current "encoding" parameter as "fallback encoding".

Not sure which solution is better. But anyway just a suggestion. Or if there is anything I can do to improve the patch, please also let me know.

BTW, Thanks for this awesome software!

### CODE REVIEW

**Comment by sanderland** [comment]
The reason for the default encoding is that the standard specifies it (https://www.red-bean.com/sgf/properties.html#CA) - I am not sure why so many chinese tools in particular ignore the standard.

I have used a library like chardet in the past, but they are often wrong on CJK encodings because there are so many of them and sgf files tend to have little text. It also adds a dependency.

@yishn how does sabaki handle this?

**Comment by xiaoyifang** [comment]
use UTF-8(international) as fallback encoding may be a way.
when load file ,parse error,try load with UTF-8.(as least an international encoding). 
also change file encoding is an easy work compared to change sgf content.

btw.
following is the lizzie's encoding detection. may be help.
https://github.com/featurecat/lizzie/blob/ebeb4469ebf7b28d6b4ca34006cb3f7059fdc9c3/src/main/java/featurecat/lizzie/rules/SGFParser.java#L25

**Comment by sanderland** [comment]
I see lizzie also ignores the standard and uses utf8 or chardet. Maybe the standard is too old. @Carton do you want to have a crack at reproducing this?

**Comment by Carton** [comment]
I found that Sabaki implements its own SGF parser, which uses js version of chardet to detect first 300 characters of SGF file. So I'll try to go with the "chardet" direction and try to implement similar detection method as Sabaki.

**Comment by Carton** [comment]
Hi @sanderland,
I've updated my previous patch to use "chardet" solution. I did not fully copy the logic in Sabaki sgf and only use a simplified logic. The only issue I can think of is that SGF files in unicode ucs-2 etc. encoding may not be correctly detected. But I think that kind of file format should be very rare.

Also added some encoding conversion logic to workaround chardet detetion issues for GB2312 related encodings (inspired by Lizzie sgf parsing code).

Feel free to modify it if you need to.

File: katrain/core/sgf_parser.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import copy
+import chardet
```
Comment (sanderland) [inline katrain/core/sgf_parser.py:2]: this is an external package, right? please add it to setup.py/requirements.txt

File: katrain/core/sgf_parser.py [hunk 2]
```diff
@@ -385,7 +386,10 @@ def parse_file(cls, filename, encoding=None) -> SGFNode:
                     if match:
                         encoding = match[1].decode("ascii", errors="ignore")
                     else:
-                        encoding = cls.DEFAULT_ENCODING
+                        encoding = chardet.detect(bin_contents[:300])['encoding']
```
Comment (sanderland) [inline katrain/core/sgf_parser.py:389]: why 300? is it too slow otherwise?

**Review by sanderland — COMMENTED** [review]
few questions

File: katrain/core/sgf_parser.py (hunk 2 above)
Comment (xiaoyifang) [reply katrain/core/sgf_parser.py:389]: https://pypi.org/project/cchardet/
cchardet seems to offer a fast character detection function .1000x faster than chardet.

Comment (sanderland) [reply katrain/core/sgf_parser.py:389]: @xiaoyifang I'm a bit hesitant to include more things that need compiling as they can give problems with os compatibility.

Comment (Carton) [reply katrain/core/sgf_parser.py:389]: @sanderland this first 300 bytes choice is copied from Sabaki sgf parser. My simple timeit test shows that detecting a typical 2KB SGF file on my PC is ~12 ms while detecting only first 300 bytes is ~ 3ms. The performance diff is negligible for typical SGF file. But based on my test, since the "encoded" words mainly appear in the beginning of SGF file, chardet shows exact same "confidence" detecting full or only first 300 bytes.

So in summary I think that optimization should be OK.

File: katrain/core/sgf_parser.py (hunk 1 above)
Comment (Carton) [reply katrain/core/sgf_parser.py:2]: Uploaded a new patch adding it to setup.py (not sure I added it the right way, though). My local "pip install ." works.
It's already listed in requirements.txt.

**Comment by sanderland** [comment]
Thanks!
